README.md
![IMA.js logo](https://imajs.io/img/imajs-logo.png)

# IMA.js

[![Build Status](https://travis-ci.com/seznam/ima.svg?branch=master)](https://travis-ci.com/seznam/ima) [![dependencies Status](https://david-dm.org/seznam/ima/status.svg)](https://david-dm.org/seznam/ima)
[![Known Vulnerabilities](https://snyk.io/test/npm/ima/badge.svg)](https://snyk.io/test/npm/ima)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)

The IMA.js is an application development stack for developing isomorphic
applications written in pure JavaScript and React.

## Why we use IMA.js and you should too?

Here at [Seznam.cz](https://www.seznam.cz), development of a frontend application comes with many checkboxes that need to be ticked off before the project goes public. Mainly because of a diverse audience and a challenging product requirements.

In order to **not** reinvent the wheel on every project and to address all of these problems (checkboxes), we have created the IMA.js framework. Here are a few outlines that we're most proud of:

- [X] **Isomorphic** - application logic is first executed at the server-side, generates the page markup, and when the application logic is executed at the client-side it automatically binds to the server-generated markup and acts like a single-page application (or a multi-page application if the client does not support JavaScript). This allows for fast load times, out-of-the-box support for web crawlers and greater overall user experience (or UX for short).
- [X] **React compatible** - IMA.js Views extend the React Component and are in tight cooperation with our Controllers. That means you can use the full magic of React v16 without loosing anything.
- [X] **Production ready** - there's no need for additional setup or configuration. IMA.js uses evironment-specific configurations from the start.
- [X] **Battle tested** - IMA.js is used on various projects across Seznam.cz. Some of them pushing the limits of what a frontend application can do.

## Documentation

We have prepared a complex tutorial for you:
[Your first IMA.js application](https://imajs.io/tutorial/introduction).
This tutorial covers the basics of creating isomorphic web applications using
IMA.js, but you will encounter some more advanced concepts in there as well.

For a more in-depth information about the IMA.js see a [full documentation](https://imajs.io/docs) and more on [imajs.io](https://imajs.io).

## Main IMA.js parts
- **core** - it contains base classes and common classes for every day work which allows you server side rendering and hydrating application in browser.
- **server** - it contains methods which allow you  to connect IMA application with the express framework.
- **gulp-tasks** - IMA.js usees gulp for bundling and automatization. There are prepared common tasks and a base configuration.

## Plugins
See the [ima-plugins](https://github.com/seznam/IMA.js-plugins) repository for available IMA.js plugins.

## Getting Started
Initialize application skeleton with single command

```bash
npx create-ima-app
# or if you are using yarn
yarn create ima-app
```

and start the development!

```bash
cd create-ima-app
npm run dev
# or if you are using yarn
yarn dev
```

## Contributing
  
Contribute to this project via [Pull-Requests](https://github.com/seznam/ima/pulls).

We are using [changesets](https://github.com/changesets/changesets/blob/main/docs/common-questions.md) for version management. To simplify the changeset creation process, you can use `npm run changeset` command. It opens an interactive interface, which should help you with changeset file composition.


### Set up the dev environment

Our dev stack expects `node>=18` and `npm>=7`.

To set up, simply run `npm install` from the repo root.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [17.15.2](https://github.com/seznam/ima/compare/v17.15.1...v17.15.2) (2022-08-10)

**Note:** Version bump only for package ima





## [17.15.1](https://github.com/seznam/ima/compare/v17.15.0...v17.15.1) (2022-05-26)

**Note:** Version bump only for package ima





# [17.15.0](https://github.com/seznam/ima/compare/v17.14.0...v17.15.0) (2022-05-20)


### Features

* scoped client window listeners ([#199](https://github.com/seznam/ima/issues/199)) ([acef240](https://github.com/seznam/ima/commit/acef24026116fff6e11999cc36c914c39b501f73))





# [17.14.0](https://github.com/seznam/ima/compare/v17.13.0...v17.14.0) (2022-04-20)


### Bug Fixes

* 🐛 Removed broken rgba style rules and disabled stylelint rule ([81e9004](https://github.com/seznam/ima/commit/81e9004dee4ee67aa61a4b408abdda7f1f00040d))


### Features

* 🎸 Changed regex in Dictionary has() function + added test ([3014c82](https://github.com/seznam/ima/commit/3014c827b0b63a5e201b34b1c7688d27214f123c))





# [17.13.0](https://github.com/seznam/ima/compare/v17.12.3...v17.13.0) (2022-04-06)


### Bug Fixes

* 🐛 Added devtools-scripts missing clone helper destructing ([06b0f9a](https://github.com/seznam/ima/commit/06b0f9a7022807be51d352d19b620fd698a3baf8))


### Features

* 🎸 new Router method getRouteHandlers ([d5312f4](https://github.com/seznam/ima/commit/d5312f4d94ca1d34d77cc0ae9d62628369484bfc))





## [17.12.3](https://github.com/seznam/ima/compare/v17.12.2...v17.12.3) (2022-03-10)


### Bug Fixes

* 🐛 always parse cookies, remove deleted cookies ([7a6dfc3](https://github.com/seznam/ima/commit/7a6dfc3418877a26d21136a3351396164a655377))
* 🐛 do not overwrite existing cookie options ([76223fc](https://github.com/seznam/ima/commit/76223fc16bd1398e7aff10334612d919a79d6d34))





## [17.12.2](https://github.com/seznam/ima/compare/v17.12.1...v17.12.2) (2022-02-07)


### Bug Fixes

* 🐛 freeze @babel/core and @babel/helpers ([4957363](https://github.com/seznam/ima/commit/49573634f7c518a638bbc9888187febc7b785fc5))





## [17.12.1](https://github.com/seznam/ima/compare/v17.12.0...v17.12.1) (2021-12-09)


### Bug Fixes

* 🐛 Clear partialState when promise was rejected ([509daaa](https://github.com/seznam/ima/commit/509daaaee5b9c98b8abbb0099e6f50cd1b1517d2))
* 🐛 HttpProxy no longer sends empty Content-Type headers ([49489b0](https://github.com/seznam/ima/commit/49489b0d26dec404ecc7351f7e0ec00a1d190fd3))





# [17.12.0](https://github.com/seznam/ima/compare/v17.11.3...v17.12.0) (2021-11-01)


### Features

* 🎸 Enable setting headers for redirects ([#141](https://github.com/seznam/ima/issues/141)) ([da0cb80](https://github.com/seznam/ima/commit/da0cb80d8719cd6b8811ceb62f7eb1e526e49193))





## [17.11.3](https://github.com/seznam/ima/compare/v17.11.2...v17.11.3) (2021-08-27)

**Note:** Version bump only for package ima





## [17.11.2](https://github.com/seznam/ima/compare/v17.11.1...v17.11.2) (2021-07-16)


### Bug Fixes

* 🐛 passing at least query params to error route ([#126](https://github.com/seznam/ima/issues/126)) ([6b1224e](https://github.com/seznam/ima/commit/6b1224edd28d4ff8e853d46e46ac2b9b4160da2e))





## [17.11.1](https://github.com/seznam/ima/compare/v17.11.0...v17.11.1) (2021-07-13)


### Bug Fixes

* 🐛 passing original params to not-found and error route ([#122](https://github.com/seznam/ima/issues/122)) ([131fb7f](https://github.com/seznam/ima/commit/131fb7fdb4d239c1efe49b9254ca1585e1b90315))





# [17.11.0](https://github.com/seznam/ima/compare/v17.10.0...v17.11.0) (2021-06-12)


### Bug Fixes

* 🐛 errorHandler middleware where app is next function ([37b5d51](https://github.com/seznam/ima/commit/37b5d5179ee2dea7a73498eb70141bbab43d9053))
* remove proxy limit in production ([81c6084](https://github.com/seznam/ima/commit/81c6084aaf80587fe0f1e007de485b980cde643f))


### Features

* 🎸 add experimental batch resolving promises ([c3704f0](https://github.com/seznam/ima/commit/c3704f06e768ca40b63bc9aa831aef747717e1dc))
* 🎸 GenericError can modify action and options params ([305436b](https://github.com/seznam/ima/commit/305436bb8b91d251e1eb9fc6859b226ece23d3f3))


### Performance Improvements

* ⚡️ batch mode keep one free frame between commits ([fa57f8c](https://github.com/seznam/ima/commit/fa57f8c13f28989a71ac7cfc98c107569d80fbbb))





# [17.10.0](https://github.com/seznam/ima/compare/v17.9.0...v17.10.0) (2021-03-30)


### Features

* 🎸 Added support for router middlewares ([#103](https://github.com/seznam/ima/issues/103)) ([d8c6d84](https://github.com/seznam/ima/commit/d8c6d845109ceba3665c0e612f5f277037a9485c))





# [17.9.0](https://github.com/seznam/ima/compare/v17.8.1...v17.9.0) (2021-03-21)


### Bug Fixes

* 🐛 assign() lefts RegExp objects intact ([3a0d52e](https://github.com/seznam/ima/commit/3a0d52e731733df07dc5bfa58db3ed087a1c4869))


### Features

* 🎸 Static and Dynamic Routes ([#98](https://github.com/seznam/ima/issues/98)) ([920b8a7](https://github.com/seznam/ima/commit/920b8a72ded69cacfa2c480020ef710bbb9ccc6a))





## [17.8.1](https://github.com/seznam/ima/compare/v17.8.0...v17.8.1) (2021-02-17)


### Bug Fixes

* 🐛 extension update method receives full state ([#102](https://github.com/seznam/ima/issues/102)) ([5c163e7](https://github.com/seznam/ima/commit/5c163e7268a46bcc7087313c58cca7097e6768c7)), closes [#101](https://github.com/seznam/ima/issues/101)





# [17.8.0](https://github.com/seznam/ima/compare/v17.7.10...v17.8.0) (2021-02-03)


### Features

* 🎸 added getTransactionStatePatches method ([5082fc1](https://github.com/seznam/ima/commit/5082fc11e03dd36abbe5793f8bbd7c2c72c3131e))
* 🎸 state transactions ([#99](https://github.com/seznam/ima/issues/99)) ([ad6237e](https://github.com/seznam/ima/commit/ad6237eabeed5b58da6e67032fc6effb33f16053))





## [17.7.10](https://github.com/seznam/ima/compare/v17.7.9...v17.7.10) (2021-01-07)



.eslintrc.js
module.exports = {
  root: true,
  ignorePatterns: [
    '**/node_modules/**',
    '**/dist/**',
    '**/build/**',
    '**/docs/**',
    '**/.turbo/**',
    '**/.docusaurus/**',
    '**/coverage/**',
    'packages/create-ima-app/examples/todos/assets/**',
  ],
  extends: [
    'eslint:recommended',
    'plugin:react/recommended',
    'plugin:react/jsx-runtime',
    'plugin:jest/recommended',
    'plugin:jest/style',
    'plugin:prettier/recommended',
  ],
  rules: {
    'no-console': [
      'error',
      {
        allow: ['warn', 'error'],
      },
    ],

    // Prettier
    'prettier/prettier': [
      'error',
      {
        singleQuote: true,
        semi: true,
        trailingComma: 'es5',
        jsxSingleQuote: true,
        bracketSameLine: false,
        arrowParens: 'avoid',
      },
    ],

    // Jest plugin overrides
    'jest/valid-title': 'off',
    'jest/no-done-callback': 'warn',
    'jest/no-disabled-tests': 'warn',
    'jest/no-conditional-expect': 'warn',
    'jest/prefer-expect-resolves': 'warn',
    'jest/prefer-lowercase-title': [
      'warn',
      {
        ignore: ['describe'],
      },
    ],
    // Remove when migrated to jest >=27
    'jest/no-jasmine-globals': 'off',

    // React plugin overrides
    'react/prop-types': 'off',
  },
  settings: {
    react: {
      version: '17',
    },
  },
  parser: '@babel/eslint-parser',
  parserOptions: {
    sourceType: 'module',
    requireConfigFile: false,
    babelOptions: {
      presets: ['@babel/preset-react'],
    },
  },
  env: {
    browser: true,
    node: true,
    es2022: true,
  },
  globals: {
    $Debug: true,
    $IMA: true,
    using: true,
    extend: true,
    spyOn: true,
  },
  overrides: [
    // TODO IMA@18 Enable repo-wide when merged to master
    // Import plugin
    {
      files: [
        'website/**',
        'packages/cli/**',
        'packages/devtools/**',
        'packages/devtools-scripts/**',
        'packages/hmr-client/**',
        'packages/error-overlay/**',
        'packages/dev-utils/**',
      ],
      extends: ['plugin:import/recommended'],
      rules: {
        'import/no-unresolved': [
          'warn',
          {
            ignore: [
              '^@\\/', // ignore @/* aliases
              '@(docusaurus|theme)',
            ],
          },
        ],
        'import/order': [
          'error',
          {
            groups: ['builtin', 'external', 'internal'],
            pathGroups: [
              {
                pattern: '{preact|react|svelte|docusaurus|theme}{/**,**}',
                group: 'external',
                position: 'before',
              },
              {
                pattern: '@/**',
                group: 'internal',
                position: 'after',
              },
              {
                pattern: '*.{css,less,json,html,txt,csv,png,jpg,svg}',
                group: 'object',
                patternOptions: { matchBase: true },
                position: 'after',
              },
            ],
            'newlines-between': 'always',
            alphabetize: {
              order: 'asc',
              caseInsensitive: true,
            },
          },
        ],
      },
      settings: {
        'import/resolver': {
          node: {
            extensions: ['.js', '.ts', '.jsx', '.tsx', '.mjs', '.json'],
          },
        },
      },
    },
    // Typescript support
    {
      files: ['**/*.{ts,tsx}'],
      parserOptions: {
        tsconfigRootDir: __dirname,
        project: './tsconfig.json',
      },
      extends: ['plugin:@typescript-eslint/recommended'],
      rules: {
        '@typescript-eslint/ban-ts-comment': [
          'error',
          { 'ts-expect-error': 'allow-with-description' },
        ],
        '@typescript-eslint/no-unused-vars': 'error',
        '@typescript-eslint/no-namespace': [
          'error',
          { allowDeclarations: true },
        ],
      },
    },
    // Website/docs overrides
    {
      files: ['website/**'],
      rules: {
        'react/react-in-jsx-scope': 'error',
        'react/jsx-uses-react': 'error',
      },
    },
    // Other overrides
    {
      files: [
        'website/scripts/**',
        'packages/cli/**',
        'packages/dev-utils/**',
        'packages/create-ima-app/**',
      ],
      rules: {
        'no-console': 'off',
      },
    },
    {
      files: ['packages/devtools/**', 'packages/create-ima-app/**'],
      rules: {
        'no-unused-vars': 'warn',
      },
      globals: {
        chrome: true,
        FB: true,
      },
    },
  ],
};
